crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,582 @@
+            let buffer_slice = input_buffer.drain(..len_write).collect::<Vec<u8>>();
+
+            desc_chain
+                .memory()
+                .write_slice(&buffer_slice, desc.addr())
+                .map_err(Error::WriteToGuestFailed)?;
+
+            // Should not overflow since "input_buffer" can hold a maximum of MAX_CAPACITY elements.
+            sent_bytes += len_write;
+
+            if input_buffer.is_empty() {
+                break;
+            }
+        }
+
+        Ok(sent_bytes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use virtio_queue::defs::VIRTQ_DESC_F_WRITE;
+    use virtio_queue::mock::MockSplitQueue;
+    use virtio_queue::Descriptor;
+    use vm_memory::{GuestAddress, GuestMemoryMmap};
+
+    // Used to check if two consoles are of the same size and have the same data in `input_buffer`.
+    impl<T> PartialEq for Console<T>
+    where
+        T: Write,
+    {
+        fn eq(&self, other: &Self) -> bool {
+            let input_buffer_self = self.input_buffer.lock().unwrap();
+            let input_buffer_other = other.input_buffer.lock().unwrap();
+            if self.capacity != other.capacity {
+                return false;
+            }
+            if input_buffer_self.len() != input_buffer_other.len() {
+                return false;
+            }
+            for i in 0..input_buffer_self.len() {
+                if input_buffer_self[i] != input_buffer_other[i] {
+                    return false;
+                }
+            }
+            true
+        }
+    }
+
+    impl PartialEq for Error {
+        fn eq(&self, other: &Self) -> bool {
+            use self::Error::*;
+            match (self, other) {
+                (WriteToGuestFailed(ref e), WriteToGuestFailed(ref other_e)) => {
+                    format!("{}", e).eq(&format!("{}", other_e))
+                }
+                (WriteToOutputFailed(ref e), WriteToOutputFailed(ref other_e)) => {
+                    format!("{}", e).eq(&format!("{}", other_e))
+                }
+                (ChainLengthOverflow, ChainLengthOverflow) => true,
+                (BufferCapacityExceeded, BufferCapacityExceeded) => true,
+                (UnexpectedReadOnlyDescriptor, UnexpectedReadOnlyDescriptor) => true,
+                (UnexpectedWriteOnlyDescriptor, UnexpectedWriteOnlyDescriptor) => true,
+                (InvalidBufferCapacity, InvalidBufferCapacity) => true,
+                _ => false,
+            }
+        }
+    }
+
+    const INPUT_VALUE: u8 = 42;
+    const INPUT_SIZE: usize = 256;
+
+    #[test]
+    fn test_default_and_partial_eq() {
+        // Default console
+        let console1 = Console::default();
+        assert_eq!(console1.capacity, DEFAULT_CAPACITY);
+
+        let console2 = Console::default();
+        assert_eq!(console2.capacity, DEFAULT_CAPACITY);
+
+        // Same size, no data
+        assert!(console1.eq(&console2));
```

Comment:
assert_eq! and assert_ne! are using the PartialEq implementation, so it would be better to use the idiomatic way to compare values instead of `assert!`s, you can find more details about this [here](https://doc.rust-lang.org/book/ch11-01-writing-tests.html#testing-equality-with-the-assert_eq-and-assert_ne-macros). But since these tests will probably disappear, it was more of an FYI.